### PR TITLE
Add a Graph NN example 

### PR DESCRIPTION
**Pull request by ksalama**
Hello @fchollet - I have finally implemented my first Graph Neural Network (GNN) model from scratch in this example. The task is to predict a research paper topic given its citations network.

I tried to provide a simplified and somehow efficient implementation for a GNN, but of course it's not optimal and doesn't cover all the task of graph learnings. I mentioned that the purpose is to provide an understanding of how GNNs work, while suggesting using specialized frameworks in research and production.

So, I am expecting a lot of review on this one (but I hope you find it useful interesting :-))

Cheers

### CODE REVIEW

File: examples/graph/gnn_citations.py [hunk 1]
```diff
@@ -0,0 +1,705 @@
...
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2021/05/30
+Last modified: 2021/05/30
+Description: Implementing a graph neural network model for predicting the topic of a paper given its citations.
+"""
+
+"""
+## Introduction
+
+Many datasets in various machine learning (ML) application have structural relationships
```
Comment (s-mrb) [inline examples/graph/gnn_citations.py:12]: `applications`

File: examples/graph/gnn_citations.py [hunk 2]
```diff
@@ -0,0 +1,705 @@
...
+between their entities, which can be represented as graphs. Such application includes
```
Comment (s-mrb) [inline examples/graph/gnn_citations.py:13]: typo, missed `s` in `applications`

File: examples/graph/gnn_citations.py [hunk 3]
```diff
@@ -0,0 +1,705 @@
...
+column_names = feature_names + ["subject"]
+papers = pd.read_csv(
+    os.path.join(data_dir, "cora.content"), sep="\t", header=None, names=column_names
+)
+num_features = papers.shape[-1] - 2
+print("Papers shape:", papers.shape)
+
+"""
+Now we display a sample of the `papers` DataFrame. The DataFrame includes the `paper_id`
+and the `subject` columns, as well as 1,433 binary column representing whether a term exists
```
Comment (s-mrb) [inline examples/graph/gnn_citations.py:104]: it would be more clear if you mention 1433 as number of rows

File: examples/graph/gnn_citations.py [hunk 4]
```diff
@@ -0,0 +1,705 @@
...
+class_values = sorted(papers["subject"].unique())
+class_idx = {name: id for id, name in enumerate(class_values)}
+
+papers["paper_id"] = papers["paper_id"].apply(lambda name: paper_idx[name])
+citations["source"] = citations["source"].apply(lambda name: paper_idx[name])
+citations["target"] = citations["target"].apply(lambda name: paper_idx[name])
+papers["subject"] = papers["subject"].apply(lambda value: class_idx[value])
+
+"""
+Now let's visualize the citation graph. Each node in the graph represent a paper,
```
Comment (s-mrb) [inline examples/graph/gnn_citations.py:132]: `represents`

File: examples/graph/gnn_citations.py [hunk 5]
```diff
@@ -0,0 +1,705 @@
...
+def display_class_probabilities(probabilities):
+    for instance_idx, probs in enumerate(probabilities):
+        print(f"Instance {instance_idx + 1}:")
+        for class_idx, prob in enumerate(probs):
+            print(f"- {class_values[class_idx]}: {round(prob * 100, 2)}%")
+        print("")
+
+
+"""
+Now we show the baseline model predictions given these randomly generated instance.
```
Comment (s-mrb) [inline examples/graph/gnn_citations.py:344]: `instances`

File: examples/graph/gnn_citations.py [hunk 6]
```diff
@@ -0,0 +1,705 @@
...
+## Build a Graph Neural Network Model
+
+### Prepare the data for the graph model
+
+Preparing and loading the graphs data into the model for training is the most challenging
+part in GNN models, which is addressed in different ways by the specialised libraries.
+In this example, we show a simple approach for preparing and using graph data that is suitable
+if your dataset consists of a single graph that fits entirely in memory.
+
+We graph data is represented by the `graph_info` tuple, which consists of the following
```
Comment (s-mrb) [inline examples/graph/gnn_citations.py:362]: Typo i guess, 

>`We` graph data

File: examples/graph/gnn_citations.py [hunk 7]
```diff
@@ -0,0 +1,705 @@
...
+three elements:
+
+1. `node_features`: This is a `[num_nodes, num_features]` NumPy array that includes the
+node features. In this dataset, the nodes are the papers, and the `node_features` are the
+word-presence binary vectors of each paper.
+2. `edges`:  This is `[num_edges, num_edges]` NumPy array representing a sparse
+[adjacency matrix](https://en.wikipedia.org/wiki/Adjacency_matrix#:~:text=In%20graph%20theory%20and%20computer,with%20zeros%20on%20its%20diagonal.)
+of the links between the nodes. In this example, the links are the citations between the papers.
+3. `edge_weights` (optional): This is a `[num_edges]` NumPy array that includes the edge weights.
```
Comment (s-mrb) [inline examples/graph/gnn_citations.py:371]: I got it its not used in this example but what edge weights would be otherwise?

File: examples/graph/gnn_citations.py [hunk 8]
```diff
@@ -0,0 +1,705 @@
...
+1. **Prepare**: The input node representations are processed using FFN. You can simplify
+the processing by only applying linear transformation to the representations.
+2. **Aggregate**: The node representations of each neighbours of each node are aggregated with
+respect to the `edge_weigts` using a *permutation invariant* pooling operation, such as *sum*, *mean*, and *max*,
+to prepare a single message for each node.
+3. **Update**: The `[num_nodes, representation_dim]` node representations and the
+`[num_nodes, representation_dim]` neighbour messages are combined (concatenated or added)
+and processed using FFN to produce the new state of the node representations (embeddings).
+
+The technique implemented used ideas from [Graph Convolutional Networks](https://arxiv.org/abs/1609.02907),
```
Comment (s-mrb) [inline examples/graph/gnn_citations.py:404]: `used`  -> `use`

File: examples/graph/gnn_citations.py [hunk 9]
```diff
@@ -0,0 +1,705 @@
...
+### Implement a graph neural network node classifier
+
+The GNN classification model follows the [Design Space for Graph Neural Networks](https://arxiv.org/abs/2011.08843) approach,
+as follows:
+
+1. Apply preprocessing using FFN to the node features to generate initial node representations.
+2. Apply one or more graph convolutional layer, with skip connections,  to the node representation
+to produce node embeddings.
+3. Apply post-processing using FFN to the node embeddings to generat the final node embeddings.
+4. Feed the the node embeddings in a Softmax layer to predict the node class.
```
Comment (s-mrb) [inline examples/graph/gnn_citations.py:506]: `the the`

File: examples/graph/gnn_citations.py [hunk 10]
```diff
@@ -0,0 +1,705 @@
...
+"""
+Let's plot the learning curves
+"""
+
+display_learning_curves(history)
+
+"""
+Now we evaluate the GNN model on the test data split.
+The results may vary depending on the training sample, however the GNN model always outperforms
+the the baseline model in terms of the test accuracy."""
```
Comment (s-mrb) [inline examples/graph/gnn_citations.py:647]: `the the`

**Review by s-mrb — COMMENTED** [review]
It's a great example!

File: examples/graph/gnn_citations.py (hunk 7 above)
Comment (ksalama) [reply examples/graph/gnn_citations.py]: OK updated

**Comment by ksalama** [comment]
Thanks @s-mrb for your comments. I have addressed them.

File: examples/graph/gnn_citations.py [hunk 11]
```diff
@@ -0,0 +1,728 @@
...
+
+1. The `cora.cites` includes the citation records with two columns:
+`cited_paper_id` (target) and `citing_paper_id` (source).
+2. The `cora.content` includes the paper content records with 1,435 columns:
+`paper_id`, `subject`, and 1,433 binary features.
+
+Let's download the dataset.
+"""
+
+urlretrieve("https://linqs-data.soe.ucsc.edu/public/lbc/cora.tgz", "cora.tgz")
```
Comment (fchollet) [inline examples/graph/gnn_citations.py:63]: Have you tried using `keras.utils.get_file(url, extract=True)`?

File: examples/graph/gnn_citations.py [hunk 12]
```diff
@@ -0,0 +1,728 @@
...
+
+"""
+## Implement Train and Evaluate Experiment
+"""
+
+hidden_units = [32, 32]
+learning_rate = 0.01
+dropout_rate = 0.5
+num_epochs = 300
+batch_size = 256
```
Comment (fchollet) [inline examples/graph/gnn_citations.py:179]: Nice to see explicit & readable variable names!

File: examples/graph/gnn_citations.py [hunk 13]
```diff
@@ -0,0 +1,728 @@
...
+"""
+
+
+def create_ffn(hidden_units, dropout_rate, name=None):
+    fnn_layers = []
+
+    for units in hidden_units:
+        fnn_layers.append(layers.BatchNormalization())
+        fnn_layers.append(layers.Dropout(dropout_rate))
+        fnn_layers.append(layers.Dense(units, activation=tf.nn.gelu))
```
Comment (fchollet) [inline examples/graph/gnn_citations.py:245]: You sure you don't want to apply the activation after the BN layer?

File: examples/graph/gnn_citations.py [hunk 14]
```diff
@@ -0,0 +1,728 @@
...
+        self.edge_weights = self.edge_weights / tf.math.reduce_sum(self.edge_weights)
+
+        self.num_classes = num_classes
+        self.hidden_units = hidden_units
+        self.aggregation_type = aggregation_type
+        self.combination_type = combination_type
+        self.dropout_rate = dropout_rate
+        self.normalize = normalize
+
+    def build(self, input_shape):
```
Comment (fchollet) [inline examples/graph/gnn_citations.py:576]: Given that you don't use `input_shape`, prefer moving this to `__init__`. The general expectation is that calling `build` should create the weights of the network, which isn't the case here.

File: examples/graph/gnn_citations.py [hunk 15]
```diff
@@ -0,0 +1,728 @@
...
+        # Preprocess the node_features to produce node representations.
+        x = self.preprocess(self.node_features)
+        # Apply the first graph conv layer.
+        x1 = self.conv1((x, self.edges, self.edge_weights))
+        # Skip connection.
+        x = self.skip_connection([x1, x])
+        # Apply the second graph conv layer.
+        x2 = self.conv2((x, self.edges, self.edge_weights))
+        # Skip connection.
+        x = self.skip_connection([x2, x])
```
Comment (fchollet) [inline examples/graph/gnn_citations.py:618]: Just `+` is more readable (for all `skip_connection` calls). This isn't a Functional model.

I guess alternatively you *could* make this model a Functional model. Up to you.

File: examples/graph/gnn_citations.py [hunk 16]
```diff
@@ -0,0 +1,728 @@
...
+
+gnn_model = GNNNodeClassifier(
+    graph_info=graph_info,
+    num_classes=num_classes,
+    hidden_units=hidden_units,
+    dropout_rate=dropout_rate,
+    name="gnn_model",
+)
+
+input_node_indicies = [1, 100, 1000]
```
Comment (fchollet) [inline examples/graph/gnn_citations.py:642]: indices (search & replace, many locations are affected)

File: examples/graph/gnn_citations.py [hunk 17]
```diff
@@ -0,0 +1,728 @@
...
+        if self.combination_type == "gru":
+            node_embeddings = tf.unstack(node_embeddings, axis=1)[-1]
+
+        if self.normalize:
+            node_embeddings = tf.nn.l2_normalize(node_embeddings, axis=-1)
+        return node_embeddings
+
+    def call(self, inputs):
+        # inputs is a tuple of three elements: node_repesentations, edges, edge_weights.
+        node_repesentations, edges, edge_weights = inputs
```
Comment (fchollet) [inline examples/graph/gnn_citations.py:503]: Inputs should be documented in a docstring

File: examples/graph/gnn_citations.py [hunk 18]
```diff
@@ -0,0 +1,728 @@
...
+            self.hidden_units, self.dropout_rate, name="postprocess"
+        )
+
+        self.compute_logits = layers.Dense(units=self.num_classes, name="logits")
+
+    def call(self, input_node_indicies):
+        # Preprocess the node_features to produce node representations.
+        x = self.preprocess(self.node_features)
+        # Apply the first graph conv layer.
+        x1 = self.conv1((x, self.edges, self.edge_weights))
```
Comment (fchollet) [inline examples/graph/gnn_citations.py:612]: I don't fully understand why edges and edge_weights are part of the call arguments if they don't vary from call to call. Sounds like they should be constructor arguments (making a layer instance specific to a set of edges)

File: examples/graph/gnn_citations.py [hunk 19]
```diff
@@ -0,0 +1,728 @@
...
+    ):
+        super(GraphConvLayer, self).__init__(*args, **kwargs)
+
+        self.hidden_units = hidden_units
+        self.aggregation_type = aggregation_type
+        self.combination_type = combination_type
+        self.dropout_rate = dropout_rate
+        self.normalize = normalize
+
+    def build(self, input_shape):
```
Comment (fchollet) [inline examples/graph/gnn_citations.py:435]: Same comment about `build` -- you can do this in the constructor since you're just creating unbuilt layers and you're not using `input_shape`

File: examples/graph/gnn_citations.py [hunk 20]
```diff
@@ -0,0 +1,728 @@
...
+The technique implemented use ideas from [Graph Convolutional Networks](https://arxiv.org/abs/1609.02907),
+[GraphSage](https://arxiv.org/abs/1706.02216), [Graph Isomorphism Network](https://arxiv.org/abs/1810.00826),
+[Simple Graph Networks](https://arxiv.org/abs/1902.07153), and
+[Gated Graph Sequence Neural Networks](https://arxiv.org/abs/1511.05493).
+Two other key techniques that are not covered are [Graph Attention Networks](https://arxiv.org/abs/1710.10903)
+and [Message Passing Neural Networks](https://arxiv.org/abs/1704.01212).
+"""
+
+
+class GraphConvLayer(keras.layers.Layer):
```
Comment (fchollet) [inline examples/graph/gnn_citations.py:416]: I'm definitely not expert enough on GNNs to review the details here. Is there someone else we could ask a second opinion from?

Comment (fchollet) [inline examples/graph/gnn_citations.py:416]: `keras.layers` -> just `layers`, it's imported

**Review by fchollet — COMMENTED** [review]
A GNN will  make a super useful example. Thank you for the PR!

File: examples/graph/gnn_citations.py (hunk 11 above)
Comment (ksalama) [reply examples/graph/gnn_citations.py:63]: Done

File: examples/graph/gnn_citations.py (hunk 13 above)
Comment (ksalama) [reply examples/graph/gnn_citations.py:245]: You mean before the dropout? I think it shouldn't make much difference. I tried to follow [Design Space for Graph NN](https://arxiv.org/pdf/2011.08843.pdf) paper (see Figure 1).

File: examples/graph/gnn_citations.py (hunk 15 above)
Comment (ksalama) [reply examples/graph/gnn_citations.py:618]: The reason I used an `Add` layer is that it displays  nicely (with a layer name "skip connection") when calling `model.summary` :-D

File: examples/graph/gnn_citations.py (hunk 14 above)
Comment (ksalama) [reply examples/graph/gnn_citations.py:576]: Done

File: examples/graph/gnn_citations.py (hunk 19 above)
Comment (ksalama) [reply examples/graph/gnn_citations.py:435]: Done

File: examples/graph/gnn_citations.py (hunk 20 above)
Comment (ksalama) [reply examples/graph/gnn_citations.py:416]: Done

File: examples/graph/gnn_citations.py (hunk 17 above)
Comment (ksalama) [reply examples/graph/gnn_citations.py:504]: Done

File: examples/graph/gnn_citations.py (hunk 18 above)
Comment (ksalama) [reply examples/graph/gnn_citations.py:604]: The reason I designed the `GraphConvLayer` to take node_features, edges, and weights as `call` parameters, rather than constructor parameters to allow using this `GraphConvLayer` in other models where you can pass a "subset" of the graph (nodes and edges) or a completely different graph in each train step. This can be useful in cases where the whole graph (info) cannot fit in memory, or in cases of graph level task (e.g. (sub)graph classification). 

In this example, we only have one graph (that can  fit in memory) and we want to preform a node-level task (node classification).  That is why in this particular example edges and edge_weights don't vary from one call to another.

WDYT?

File: examples/graph/gnn_citations.py (hunk 15 above)
Comment (fchollet) [reply examples/graph/gnn_citations.py:611]: If you want to keep this a subclassed model, then we should definitely use `+` in `call` rather than an `Add` layer. `Add` only makes sense in the context of a Functional model (addition as a layer), otherwise in `call` you should just use raw TF ops since you're not constructing a graph topology.

Comment (ksalama) [reply examples/graph/gnn_citations.py:611]: Done

File: examples/graph/gnn_citations.py (hunk 20 above)
Comment (8bitmp3) [reply examples/graph/gnn_citations.py]: @fchollet We could ask @Mistobaan (contributed TF2 code to DeepMind's Graph Nets) and colleagues @alvarosg (DM's Graph Nets), @jg8610 (DM's jraph library), @theahura (cool presentation at NeurIPS'20 [Graph Mining](https://gm-neurips-2020.github.io) by Google Research) - if you or someone you know have time, can you please take a quick glance at this GNN conv layer class? Thank you 👍

Comment (theahura) [reply examples/graph/gnn_citations.py]: Just did a quick review, it looks fine to me. I only reviewed to answer the question 'does this implement a GNN' which it does, I didn't look at style so not sure if this is consistent with other examples or anything like that.

**Review by fchollet — APPROVED** [review]
Everything is looking great! Please add the generated files.

**Comment by ksalama** [comment]
@fchollet Thanks! I added the files. Cheers

**Review by fchollet — APPROVED** [review]
Thank you for the great contribution, Khalid!
